design_docs/2023_07_20_dynamic_shader_parameter_amount.md

```diff
@@ -0,0 +1,55 @@
+# Passing variables to the shader
+We need to be able to pass an arbitrary number of textures and parameters to the custom shaders. For parameters, we need to be able to pass some "header parameters" -- constants used in all videos, as well as some per-video data.
+
+### Possible solutions
+Binding runtime-sized texture arrays works on every backend (except wasm, but it doesn't matter):
+```wgsl
+@group(0) @binding(0) var textures: texture_2d_array<f32>;
+```
+
+Binding const-sized uniform/storage buffers works everywhere too. This would work approximately like this in the shaders (ofc the len can be defined in a different way, e.g. as the length of a runtime-sized texture array):
+```wgsl
+struct MyList {
+	// you declare an array with a constant, large size
+	per_video_data: array<MyStruct, 64>
+	// and then in some way specify how much of this buffer is filled
+	per_video_data_len: u32
+}
+
+@group(0) @binding(0) var<uniform> per_video_data: MyList
```

Comment:
Change names `MyList`, `per_video_data`, and `per_video_data_len` to more descriptive ones.

Reply:
I'm against doing that, since this is supposed to represent the users capabilities of defining their own structs and using them here, which I feel these names demonstrate pretty well

Reply:
I think it's not clear here how passing user-defined parameters and textures from other nodes fit together. Based on this description I understand that:
- solution 1 - Textures are passed totally separate from params, example code is only addressing passing textures. 
- soultion 2 
  - If this is user-defined I don't see where to texture is passed. Is it MyStruct? If yes is texture + params or just texture?
  - If It's user-defined how can distinguish which values in the struct are params that the user should specify and which are buffers/textures with image data? How would we construct those structs on the rust side? (this applies to all solutions after this one too)

Reply:
Yeah, this is a little bit unclear. I'll try to add a section explaining this better.

Textures are indeed passed separately from params, in a different bind group. The full set of bindings the user declares would look approximately like this:
```wgsl
@group(0) @binding(0) var textures: texture_2d_array<f32>;
@group(1) @binding(0) var<uniform> users_custom_parameters: UsersCustomStruct;
// this can include, for example, the timestamp of this frame 
@group(1) @binding(1) var<uniform> parameters_received_from_the_compositor: OurStruct;
@group(2) @binding(0) var sampler_: sampler;
```